### PR TITLE
[script.module.inputstreamhelper] 0.2.1

### PR DESCRIPTION
**Pull request by emilsvennesson**
### Description
A simple Kodi module that makes life easier for add-on developers relying on InputStream based add-ons and DRM playback. The previous pull request discussion became quite heated so I decided to close it until a solution for ARM based devices had been found. By extracting the binary from a Chrome OS recovery image, we believe that this way should fully comply with the Widevine license.

See https://github.com/emilsvennesson/script.module.inputstreamhelper/issues/1 for more details.
### Checklist:

- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [script.foo.bar] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by asciidisco** [comment]
Anything that can be done to assist in reviewing this, or to provide some more information?
I´d really would like to use this as a dependency in my plugins.

File: script.module.inputstreamhelper/lib/inputstreamhelper.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+                        busy_dialog.close()
+                        dialog.ok(self._language(30004), self._language(30005))
+
+        return False
+
+    def _widevine_eula(self):
+        """Display the Widevine EULA and prompt user to accept it."""
+        if os.path.exists(os.path.join(self._cdm_path(), config.WIDEVINE_LICENSE_FILE)):
+            license_file = os.path.join(self._cdm_path(), config.WIDEVINE_LICENSE_FILE)
+            with open(license_file, 'r') as f:
```
Comment (wsnipex) [inline script.module.inputstreamhelper/lib/inputstreamhelper.py]: please check that file size is within reasonable limits

File: script.module.inputstreamhelper/lib/inputstreamhelper.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        """Mount loop device to self._mnt_path()"""
+        dialog = xbmcgui.Dialog()
+        cmd = ['mount', '-t', 'ext2', self._loop_dev, '-o', 'ro', self._mnt_path()]
+        if os.getuid() != 0 and self._cmd_exists('sudo'):  # ask for permission to wrap cmd in sudo
+            if dialog.yesno(self._language(30001), self._language(30030), yeslabel=self._language(30029), nolabel=self._language(30028)):
+                cmd.insert(0, 'sudo')
+            else:
+                self._log('User refused to give sudo permission.')
+
+        success = self._run_cmd(cmd)
```
Comment (wsnipex) [inline script.module.inputstreamhelper/lib/inputstreamhelper.py]: please drop sudo rights here (sudo -k)

File: script.module.inputstreamhelper/lib/inputstreamhelper.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+    def _set_loop_dev(self):
+        """Set an unused loop device that's available for use."""
+        cmd = ['losetup', '-f']
+        self._loop_dev = subprocess.check_output(cmd).strip()
+        self._log('Found free loop device: {0}'.format(self._loop_dev))
+        return True
+
+    def _losetup(self, bin_path):
+        """Setup Chrome OS loop device."""
+        cmd = ['losetup', self._loop_dev, bin_path, '-o', self._parse_chromeos_offset(bin_path)]
```
[the remark this answers is not in the document]
Comment (emilsvennesson) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: @wsnipex Are you sure? I tried it on OSMC and root rights was not needed. Perhaps it is needed on 'normal' distros.

File: script.module.inputstreamhelper/lib/inputstreamhelper.py (hunk 2 above)
Comment (emilsvennesson) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: Do you mean run sudo -k after the command has been executed?

File: script.module.inputstreamhelper/lib/inputstreamhelper.py (hunk 1 above)
Comment (emilsvennesson) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: I think this should be fine. Downloading the license from the x86 files doesn't require more than a few megabytes (2 to be exact).

File: script.module.inputstreamhelper/lib/inputstreamhelper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wsnipex) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: Maybe they made it suid root on osmc.
On ubuntu x86_64, it does require root

File: script.module.inputstreamhelper/lib/inputstreamhelper.py (hunk 2 above)
Comment (wsnipex) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: yes, we really need to limit root access to the absolute minimum.
Getting by without root would be very much preferred, although I haven't found a way that works for me yet without jumping through hoops.

**Comment by emilsvennesson** [comment]
PR updated. 

Changes: https://github.com/emilsvennesson/script.module.inputstreamhelper/compare/v0.2.0...v0.2.1

File: script.module.inputstreamhelper/lib/inputstreamhelper.py (hunk 2 above)
Comment (emilsvennesson) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: Yeah, I also hoped for a way to do it without root but it's a no-go. :-/

**Comment by asciidisco** [comment]
If it helps, here´s a branch of the Netflix Plugin using this Addon: https://github.com/asciidisco/plugin.video.netflix/tree/feat/add-inputstream-helper

Tested it on OSX & Libreelec ARM installations in a "real world" scenario.

File: script.module.inputstreamhelper/lib/inputstreamhelper.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+        success = self._run_cmd(cmd, sudo=True, ask=True)
+        if success:
+            self._attached_loop_dev = True
+            return True
+        else:
+            return False
+
+    def _mnt_loop_dev(self):
+        """Mount loop device to self._mnt_path()"""
+        cmd = ['mount', '-t', 'ext2', self._loop_dev, '-o', 'ro', self._mnt_path()]
```
Comment (wsnipex) [inline script.module.inputstreamhelper/lib/inputstreamhelper.py]: did you try mount -o loop,offset=... ?
That way you could drop the extra losetup above

Comment (emilsvennesson) [reply script.module.inputstreamhelper/lib/inputstreamhelper.py]: @wsnipex Yes, but it fails with 'Invalid argument' for some reason.

**Comment by MartijnKaijser** [comment]
Once comments are handled we can get this merged

**Comment by emilsvennesson** [comment]
@MartijnKaijser Which comments are you referring to? I believe I've addressed them all except dropping losetup since I couldn't get that to work.

I'm currently abroad for 7 more weeks without access to anything but my phone so it's kind of difficult for me to do any changes right now. :-/

**Pull request merged**